File: HERENCIA/EJERCICIO11/JefePolicia.py
from abc import ABC, abstractmethod
class IPersona(ABC):
    @abstractmethod
    def mostrar_datos_personales(self):
        pass
    
class IEmpleado(ABC):
    @abstractmethod
    def calcular_salario(self):
        pass

class IPolicia(ABC):
    @abstractmethod
    def mostrar_rango(self):
        pass

class Persona:
    def __init__(self, nombre, edad):
        self.nombre = nombre
        self.edad = edad

class Empleado:
    def __init__(self, sueldo, codigo):
        self.sueldo = sueldo
        self.codigo = codigo

class Policia:
    def __init__(self, grado, unidad):
        self.grado = grado
        self.unidad = unidad

class JefePolicia(Persona, Empleado, Policia, IPersona, IEmpleado, IPolicia):
    def __init__(self, nombre, edad, sueldo, codigo, grado, unidad):
        Persona.__init__(self, nombre, edad)
        Empleado.__init__(self, sueldo, codigo)
        Policia.__init__(self, grado, unidad)
    
    def mostrar_datos_personales(self):
        return f"Nombre: {self.nombre}, Edad: {self.edad}"
    
    def calcular_salario(self):
        return self.sueldo
    
    def mostrar_rango(self):
        return f"Grado: {self.grado}, Unidad: {self.unidad}"
    
    def mostrar_datos(self):
        print(f"Jefe de Policía: {self.nombre}")
        print(f"Edad: {self.edad}")
        print(f"Código: {self.codigo}")
        print(f"Sueldo: {self.sueldo}")
        print(f"Grado: {self.grado}")
        print(f"Unidad: {self.unidad}")

jefe1 = JefePolicia("Roberto Gómez", 45, 12000, "POL-001", "Coronel", "Unidad Táctica")
jefe2 = JefePolicia("Ana Martínez", 42, 15000, "POL-002", "Coronel", "Unidad de Investigación")

print("JEFE 1:")
jefe1.mostrar_datos()
print("\nJEFE 2:")
jefe2.mostrar_datos()

if jefe1.sueldo > jefe2.sueldo:
    print(f"El jefe con mayor sueldo es: {jefe1.nombre} con un sueldo de {jefe1.sueldo}")
elif jefe2.sueldo > jefe1.sueldo:
    print(f"El jefe con mayor sueldo es: {jefe2.nombre} con un sueldo de {jefe2.sueldo}")
else:
    print(f"Ambos tienen el mismo sueldo: {jefe1.sueldo}")

if jefe1.grado == jefe2.grado:
    print(f"Ambos jefes tienen el mismo grado: {jefe1.grado}")
else:
    print(f"Los jefes tienen diferentes grados: {jefe1.nombre} es {jefe1.grado} y {jefe2.nombre} es {jefe2.grado}")

print(f"\nDatos personales de {jefe1.nombre}: {jefe1.mostrar_datos_personales()}")
print(f"Salario calculado: {jefe1.calcular_salario()}")
print(f"Rango: {jefe1.mostrar_rango()}")

File: HERENCIA/EJERCICIO11/JefePolicia.java
package EJERCICIO11;
interface IPersona {
    String mostrar_datos_personales();
}

interface IEmpleado {
    int calcular_salario();
}

interface IPolicia {
    String mostrar_rango();
}

class Persona {
    protected String nombre;
    protected int edad;

    public Persona(String nombre, int edad) {
        this.nombre = nombre;
        this.edad = edad;
    }
}

class Empleado {
    protected int sueldo;
    protected String codigo;

    public Empleado(int sueldo, String codigo) {
        this.sueldo = sueldo;
        this.codigo = codigo;
    }
}

class Policia {
    protected String grado;
    protected String unidad;

    public Policia(String grado, String unidad) {
        this.grado = grado;
        this.unidad = unidad;
    }
}

class JefePolicia extends Persona implements IPersona, IEmpleado, IPolicia {
    public int sueldo;
    public String codigo;
    public String grado;
    public String unidad;

    public JefePolicia(String nombre, int edad, int sueldo, String codigo, String grado, String unidad) {
        super(nombre, edad);
        this.sueldo = sueldo;
        this.codigo = codigo;
        this.grado = grado;
        this.unidad = unidad;
    }

    @Override
    public String mostrar_datos_personales() {
        return "Nombre: " + this.nombre + ", Edad: " + this.edad;
    }

    @Override
    public int calcular_salario() {
        return this.sueldo;
    }

    @Override
    public String mostrar_rango() {
        return "Grado: " + this.grado + ", Unidad: " + this.unidad;
    }

    public void mostrar_datos() {
        System.out.println("Jefe de Policía: " + this.nombre);
        System.out.println("Edad: " + this.edad);
        System.out.println("Código: " + this.codigo);
        System.out.println("Sueldo: " + this.sueldo);
        System.out.println("Grado: " + this.grado);
        System.out.println("Unidad: " + this.unidad);
    }
}

